File: Question/25_Reverse_Nodes_in_K-Group.cpp
//这题没到k是不要翻转的！！！
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode dummy(0);
        dummy.next = head;
        ListNode* cur = head;
        ListNode* prev = &dummy;
        for(int i = 1; cur; ++i) {
            if(i % k == 0) { //找到分段
                prev = reverseOneGroup(prev, cur->next);
                cur = prev->next;
            } else {
                cur = cur->next;
            }
        }
        return dummy.next;
    }
    ListNode* reverseOneGroup(ListNode* pre, ListNode* next) {
        ListNode *last = pre->next, *cur = last->next;
        while(cur != next) {
            last->next = cur->next;
            cur->next = pre->next;
            pre->next = cur;
            cur = last->next;
        }
        return last;
    }
};

File: Question_Java/25_Reverse_Nodes_in_K-Group.java
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prev = dummy;
        ListNode last = head;
        for (int i = 1; last != null; ++i) {
            if (i % k != 0) {
                last = last.next;
                continue;
            }
            prev = reverse(prev, last.next);
            last = prev.next;
        }
        return dummy.next;
    }
    private ListNode reverse(ListNode prev, ListNode end) {
        ListNode last = prev.next;
        ListNode cur = last.next;
        while (cur != end) {
            last.next = cur.next;
            cur.next = prev.next;
            prev.next = cur;
            cur = last.next;
        }
        return last;
    }
}